torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,280 @@
+#include <string>
+
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/DispatchKey.h>
+#include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
+
+// _c10d_functional ops and relevant runtime APIs.
+//
+// _c10d_functional is a native port of c10d_functional which currently resides
+// in python. _c10d_functional will replace c10d_functional once functionality
+// parity is verified.
+//
+// Motivations for porting c10d_functional ops to c++:
+// - Support collective ops for aot inductor and native interpreters
+// - Support multi-threaded native runtimes
+// - Unify collective codegens between inductor python and aot inductor
+//
+// Changes compared to c10d_functional:
+// - The namespace now contains in-place variants for collective calls inductor
+// decides to de-functionalize (should we put them in a different namespace?).
+// - Process group resolution now only relies on tag. Ranks are useful for the
+// compiler to determine whether two collectives using different process groups
+// are fuse-able. However, they are not required for process group resolution.
+// - Process group resolution now doesn't upsert a process group.
+
+static thread_local std::unordered_map<void*, c10::intrusive_ptr<c10d::Work>>
+    pendingWork;
+
+namespace {
+
+const std::unordered_map<std::string, c10d::ReduceOp> strToReduceOp = {
+    {"sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::SUM)},
+    {"avg", c10d::ReduceOp(c10d::ReduceOp::RedOpType::AVG)},
+    {"product", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PRODUCT)},
+    {"min", c10d::ReduceOp(c10d::ReduceOp::RedOpType::MIN)},
+    {"max", c10d::ReduceOp(c10d::ReduceOp::RedOpType::MAX)},
+    {"band", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BAND)},
+    {"bor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BOR)},
+    {"bxor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BXOR)},
+    // {"premul_sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PREMUL_SUM)},
+    {"unused", c10d::ReduceOp(c10d::ReduceOp::RedOpType::UNUSED)}};
+
+at::Tensor all_reduce_(
+    at::Tensor input,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
```

Comment:
curious how `ranks` is used? 

I don't see this used anywhere.

Reply:
`ranks` is intended to be consumed by compilers. For example, it's possible to fuse two all_reduces issued with two groups of the same membership. However, compilers can't determine the legitimacy of such fusion in absence of membership information.

The Python user facing APIs also leverages the information for lazy sub-pg initialization (I don't think we should do the same in native mode though).